### PR TITLE
feat(or filtering properties)!: FE implementation of new properties and backwards compatibility

### PR DESCRIPTION
**Pull request by liyiy**
## Changes

*Please describe. If this affects the frontend, include screenshots.*

FE implementation of new property groups (and or) filters. Use the feature flag `and-or-filtering` and check out the different insights tabs to test. 

https://user-images.githubusercontent.com/25164963/157359767-545f3d56-b45f-4e48-b69a-d1b5dae6d44a.mov

https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=7710%3A42576

*Stay up-to-date with our [coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

## How did you test this code?

*Briefly describe the steps you took.*

### CODE REVIEW

File: frontend/src/types.ts [hunk 1]
```diff
@@ -883,7 +883,7 @@ export interface FilterType {
     interval?: IntervalType
     date_from?: string | null
     date_to?: string | null
-    properties?: PropertyFilter[]
+    properties?: PropertyGroupFilter
```
Comment (liyiy) [inline frontend/src/types.ts:886]: @EDsCODE @neilkakkar Before I refactor a bunch of stuff, I'm confirming this is the change we're looking for? We will only handle that format for properties moving forwards?

Comment (EDsCODE) [reply frontend/src/types.ts:886]: the API can accept either. We still need to be backwards compatible with the properties that exist in saved insights so I dont think a complete type swapover will work. hmmm

Comment (liyiy) [reply frontend/src/types.ts:886]: yeah that's true 😬

File: frontend/src/scenes/insights/InsightTabs/FunnelTab/FunnelTab.tsx [hunk 2]
```diff
@@ -130,23 +130,24 @@ export function FunnelTab(): JSX.Element {
...
-                        TaxonomicFilterGroupType.PersonProperties,
-                        ...groupsTaxonomicTypes,
-                        TaxonomicFilterGroupType.Cohorts,
-                        TaxonomicFilterGroupType.Elements,
-                    ]}
-                    eventNames={allEventNames}
-                />
+                {filters.properties && (
+                    <PropertyGroupFilters
+                        propertyFilters={convertPropertiesToPropertyGroup(filters.properties)}
```
Comment (liyiy) [inline frontend/src/scenes/insights/InsightTabs/FunnelTab/FunnelTab.tsx:135]: We will be removing the old property filters entirely for the global filters on `trend tab, funnel tab, etc` and implement the new property group filters. Any old properties will be converted to property group formatting here

File: frontend/src/lib/utils.tsx [hunk 3]
```diff
@@ -1325,4 +1328,39 @@ export function getEventNamesForAction(actionId: string | number, allActions: Ac
...
+    return false
+}
+
+export function convertPropertiesToPropertyGroup(
+    properties: PropertyGroupFilter | AnyPropertyFilter[]
+): PropertyGroupFilter {
+    if (isPropertyGroup(properties)) {
+        return properties
+    }
+    return { type: FilterLogicalOperator.And, values: [{ type: FilterLogicalOperator.And, values: properties }] }
```
Comment (EDsCODE) [inline frontend/src/lib/utils.tsx:1346]: We may want to revisit `PropertyGroupFilter` <> `PropertyGroupFilterValue` because this doesn't need to nest further here. Should ideally just return ``
```suggestion
    return { type: FilterLogicalOperator.And, values: properties }
```

Comment (liyiy) [reply frontend/src/lib/utils.tsx:1346]: So for backwards compatible properties, can just pass in `properties` to `values` instead of the new format with `type`? It does make that part easier but introduces another "option" of formatting

Comment (EDsCODE) [reply frontend/src/lib/utils.tsx:1346]: Eh I think ignore what I'm suggesting it's unnecessary

File: frontend/src/lib/utils.tsx [hunk 4]
```diff
@@ -1325,4 +1330,40 @@ export function getEventNamesForAction(actionId: string | number, allActions: Ac
...
+}
+
+export function convertPropertyGroupToProperties(properties: PropertyGroupFilter | PropertyFilter[]): PropertyFilter[] {
+    if (isPropertyGroup(properties)) {
+        return flattenPropertyGroup(properties)
+    }
+    return properties
+}
+
+export function flattenPropertyGroup(properties: PropertyGroupFilter): PropertyFilter[] {
```
Comment (neilkakkar) [inline frontend/src/lib/utils.tsx:1361]: quick flyby: this is too naive, since `PropertyGroup.values` can be another propertyGroup.

So, recurse, like so:
https://github.com/PostHog/posthog/blob/master/posthog/models/property.py#L191

Comment (liyiy) [reply frontend/src/lib/utils.tsx:1361]: @EDsCODE

Comment (EDsCODE) [reply frontend/src/lib/utils.tsx:1361]: I think the case where a propertyGroup might just be one layer is actually more important to consider here. We shouldn't end up with scenarios where more than two layer groups are returned to the frontend as our component couldn't handle it anyway. But, for the areas that need flattening, this util should be able to handle any depth, specifically the shallow nested groups incase some simple prop group conversion gets passed here

tl;dr we should recurse mainly to handle the shallow nesting and as a side effect will handle any nesting anyway

File: frontend/src/scenes/performance/webPerformanceLogic.ts [hunk 5]
```diff
@@ -330,7 +331,11 @@ export const webPerformanceLogic = kea<webPerformanceLogicType<EventPerformanceD
     loaders: ({ values }) => ({
         pageViewEvents: {
             loadEvents: async () => {
-                const combinedProperties = [...(eventApiProps.properties || []), ...values.properties]
+                const flattenedPropertyGroup =
+                    eventApiProps.properties && isPropertyGroup(eventApiProps.properties)
```
Comment (EDsCODE) [inline frontend/src/scenes/performance/webPerformanceLogic.ts:335]: don't think you need to null check `eventApiProps.properties` before isPropertyGroup because the function itself will handle the case if it's null. This is done in a few other areas where isPropertyGroup is called too

**Review by EDsCODE — COMMENTED** [review]
~~Hmm, how did you trigger the requests when you change the filters? I just tried manual QA and seems like changes aren't triggering new query requests~~

Just as i posted this and tested again it works

**Comment by EDsCODE** [comment]
One empty group and having another non empty group causes the query to fail. We'll want to prune empty filter objects from the object

File: posthog/models/filters/mixins/property.py [hunk 6]
```diff
@@ -101,11 +101,14 @@ def _parse_property_group_list(self, prop_list: Optional[List]) -> Union[List[Pr
             return cast(List[Property], [])
         has_property_groups = False
         has_simple_properties = False
+
         for prop in prop_list:
             if "type" in prop and "values" in prop:
                 has_property_groups = True
-            else:
```
Comment (liyiy) [inline posthog/models/filters/mixins/property.py:107]: @EDsCODE Empty properties shouldn't be considered simple properties? Not sure if this covers the edge cases correctly

**Comment by EDsCODE** [comment]
@mariusandra would be great if you could have eyes on this. It changes how properties have to be handled on the frontend. There's some overhead now because we want to be backwards compatible

File: frontend/src/scenes/insights/InsightTabs/TrendTab/TrendTab.tsx [hunk 7]
```diff
@@ -114,11 +113,13 @@ export function TrendTab({ view }: TrendTabProps): JSX.Element {
                     )}
                     {filters.insight !== InsightType.LIFECYCLE && (
                         <>
-                            {featureFlags[FEATURE_FLAGS.AND_OR_FILTERING] ? (
+                            {filters.properties && (
```
Comment (EDsCODE) [inline frontend/src/scenes/insights/InsightTabs/TrendTab/TrendTab.tsx:116]: Note: we'll want to keep this feature flag before merging

File: frontend/src/scenes/trends/personsModalLogic.tsx [hunk 8]
```diff
@@ -78,14 +86,17 @@ export function parsePeopleParams(peopleParams: PeopleParamType, filters: Partia
 
     // If breakdown type is cohort, we use breakdown_value
     // If breakdown type is event, we just set another filter
+    const flattenedPropertyGroup = isPropertyGroup(params.properties)
```
Comment (EDsCODE) [inline frontend/src/scenes/trends/personsModalLogic.tsx:89]: Thoughts on moving the isPropertyGroup check into `convertPropertyGroupToProperties`? If we know at this point we need flattened properties we can just let the function determine how to handle. Will reduce repetition

File: posthog/models/filters/mixins/property.py (hunk 6 above)
Comment (EDsCODE) [reply posthog/models/filters/mixins/property.py:107]: hmm let's just filter the `prop_list` before this for loop with something like 
`prop_list = filter(lambda prop: prop, prop_list)`

It'll make this logic more clear. Any empty dict shouldn't be a either type of prop

File: frontend/src/scenes/feature-flags/featureFlagLogic.tsx [hunk 9]
```diff
@@ -84,7 +85,19 @@ export const featureFlagLogic = kea<featureFlagLogicType>({
...
+                    newFeatureFlag.filters.groups = newFeatureFlag.filters.groups.map((group) => {
+                        if (group.properties) {
+                            return {
+                                ...group,
+                                properties: convertPropertyGroupToProperties(group.properties),
+                            }
+                        }
+                        return group
+                    })
+                    return newFeatureFlag
```
Comment (mariusandra) [inline frontend/src/scenes/feature-flags/featureFlagLogic.tsx:99]: Is this the right place for this logic? When would we run `setFeatureFlag(almostARealFeatureFlagButRequiresSomeTweaking)`? Just asking, not being sure :).

However, if this is the right place, then why not embrace the functional way of doing things?

```ts
const newFeatureFlag = {
    ...featureFlag,
    filters: {
        ...featureFlag.filters,
        groups: /* etc */
    }
}
```

**Review by mariusandra — COMMENTED** [review]
I had a look since I was tagged, but I didn't go super deep. Nothing obviously erroneous at first glance, just some nits... and a suggestion to remove the `clone` package. 

The UI is a bit clunky and cramped though, though I assume I'm not reporting anything new by saying this.

image
 
image

File: posthog/models/filters/mixins/property.py (hunk 6 above)
Comment (EDsCODE) [reply posthog/models/filters/mixins/property.py:107]: If there's an empty dict, this for loop will set both flags to be true and then return the error. So the empty dict needs to be purged

Comment (neilkakkar) [reply posthog/models/filters/mixins/property.py:107]: (just seeking context here): When do we get empty dicts, that are valid when sent from the frontend, but purged here?

I'd say there shouldn't be any case where this can happen, and if it is happening, should fix it where its being sent, and keeping the API stricter?

File: frontend/src/scenes/insights/InsightTabs/TrendTab/TrendTab.tsx (hunk 7 above)
Comment (EDsCODE) [reply frontend/src/scenes/insights/InsightTabs/TrendTab/TrendTab.tsx:116]: The usual reasons of maintaining phased rollout control and easy feature shut off if things are buggy

File: posthog/models/filters/mixins/property.py (hunk 6 above)
Comment (EDsCODE) [reply posthog/models/filters/mixins/property.py:107]: I thought so too but at the moment if you click new property and don't fill it out, you'll have an empty dict. This is the case with our current property filters too, it's just handled on the backend. I'd say it should be safe on both frontend and backend

Comment (neilkakkar) [reply posthog/models/filters/mixins/property.py:107]: brrrrrrr. Okay, thanks.

File: frontend/src/scenes/feature-flags/featureFlagLogic.tsx [hunk 10]
```diff
@@ -11,12 +11,13 @@ import {
...
 import { toast } from 'react-toastify'
 import { router } from 'kea-router'
-import { deleteWithUndo } from 'lib/utils'
+import { convertPropertyGroupToProperties, deleteWithUndo } from 'lib/utils'
 import { urls } from 'scenes/urls'
 import { teamLogic } from '../teamLogic'
 import { featureFlagsLogic } from 'scenes/feature-flags/featureFlagsLogic'
 import { groupsModel } from '~/models/groupsModel'
 import { TaxonomicFilterGroupType } from 'lib/components/TaxonomicFilter/types'
+import clone from 'clone'
```
[the remark this answers is not in the document]
Comment (liyiy) [reply frontend/src/scenes/feature-flags/featureFlagLogic.tsx:20]: sanity check that this doesn't mutate objects? 😭 

```
addFilterGroup: (state) => {
                   const filterGroups = [...state.values, { type: FilterLogicalOperator.And, values: [{}] }]
                    return { ...state, values: filterGroups }
                },
                removeFilterGroup: (state, { filterGroup }) => {
                    const filteredGroups = [...state.values]
                    filteredGroups.splice(filterGroup, 1)
                    return { ...state, values: filteredGroups }
                },```

File: frontend/src/scenes/feature-flags/featureFlagLogic.tsx (hunk 9 above)
Comment (liyiy) [reply frontend/src/scenes/feature-flags/featureFlagLogic.tsx]: I am not sure actually, I just need to make sure wherever we're setting a feature flag that if the groups in that feature flag come back as a property group, it needs to be converted to flat properties. And this should be updated now without the clone library 😄

Comment (neilkakkar) [reply frontend/src/scenes/feature-flags/featureFlagLogic.tsx]: Hmmmm, I think feature flag filters are stored completely separately, and never touch the `Filter` object for queries, so backend can't parse, and won't deal with new property groups here.

Same in every other model really, that has a `filters` key, but no corresponding `Filter` object generation (which uses the PropertyMixin) . Ouch, this is a bit confusing.

Basically implies:

## No new properties

1. Feature Flags
2. Actions
3. ???

## New Properties

1. Entities
2. Insight filters
3. by extension, dashboard filters^
4. Experimentation filters
5. ???

Comment (neilkakkar) [reply frontend/src/scenes/feature-flags/featureFlagLogic.tsx]: though @EDsCODE should confirm too, I might be missing something 😅

Comment (EDsCODE) [reply frontend/src/scenes/feature-flags/featureFlagLogic.tsx]: is there a downside in just applying the conversion logic uniformly? Even if feature flag logic doesn't need to account for new filtering format, it might be less headache to just treat it the same as properties elsewhere.

Comment (liyiy) [reply frontend/src/scenes/feature-flags/featureFlagLogic.tsx]: confused, what's the consensus on this? I vote for just applying the conversion logic since that's what the typing is yelling at me for

**Comment by clarkus** [comment]
Just a quick update to share the latest filter iteration at https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=7710%3A42576

Filters flow

File: frontend/src/scenes/trends/personsModalLogic.tsx (hunk 8 above)
Comment (liyiy) [reply frontend/src/scenes/trends/personsModalLogic.tsx]: done

**Comment by mariusandra** [comment]
Any reason we don't want to move the "filter out test users" toggle below the "add filter group" button? It feels weird if that toggle is sandwiched between between the list and the button that operates on the list.

**Comment by liyiy** [comment]
> Any reason we don't want to move the "filter out test users" toggle below the "add filter group" button? It feels weird if that toggle is sandwiched between between the list and the button that operates on the list.

Either way is fine with me, though I guess having the "filter test users" toggle below add filter group is more similar to how it currently works, I'm impartial here, @clarkus ?

File: cypress/integration/retention.js [hunk 11]
```diff
@@ -11,6 +11,9 @@ describe('Retention', () => {
         // stable. This is to try and avoid an issue where after selecting a
         // filter property, the results section would be blank
         cy.get('[data-attr=retention-table').should('exist')
+        // tests for new property group filters
+        // cy.get('[data-attr=insight-retention-add-filter-group]').click()
+        // cy.get('[data-attr=property-select-toggle-0]').click()
```
Comment (EDsCODE) [inline cypress/integration/retention.js:16]: not needed anymore?

**Comment by liyiy** [comment]
note: layout issues that don't directly impact functionality should be scoped into a follow up PR 😄

**Review by EDsCODE — COMMENTED** [review]
Bug: If you save the insight, then go to edit it, previously set OR filter will become "and" and end up in this out of sync state
Screen Shot 2022-03-09 at 10 52 10 AM

File: posthog/models/filters/mixins/property.py (hunk 6 above)
Comment (EDsCODE) [reply posthog/models/filters/mixins/property.py:107]: @liyiy thoughts on my suggestion?

File: cypress/integration/retention.js (hunk 11 above)
Comment (liyiy) [reply cypress/integration/retention.js:16]: it'll be needed when we remove the feature flags 😁

Comment (EDsCODE) [reply cypress/integration/retention.js:16]: Ah, true. Let's put a TODO there so it's noted

Comment (liyiy) [reply cypress/integration/retention.js:16]: ok!!

**Review by EDsCODE — APPROVED** [review]
Happy to merge this to start internal testing!

File: frontend/src/lib/components/PropertyGroupFilters/propertyGroupFilterLogic.ts [hunk 12]
```diff
@@ -38,41 +41,32 @@ export const propertyGroupFilterLogic = kea<propertyGroupFilterLogicType>({
                             ],
                         }
                     }
-                    const groupsCopy = clone(state)
-                    groupsCopy.values.push({ type: FilterLogicalOperator.And, values: [{}] })
+                    const filterGroups = [...state.values, { type: FilterLogicalOperator.And, values: [{}] }]
```
Comment (alexkim205) [inline frontend/src/lib/components/PropertyGroupFilters/propertyGroupFilterLogic.ts:44]: nice, not sure why we had an external library for this lol

File: frontend/src/scenes/cohorts/cohortLogic.ts [hunk 13]
```diff
@@ -31,7 +32,17 @@ function determineMatchType(group: Partial<CohortGroupType>): MatchType {
 function processCohortOnSet(cohort: CohortType): CohortType {
     if (cohort.groups) {
         cohort.groups = cohort.groups.map((group) => addLocalCohortGroupId(group))
+        cohort.groups = cohort.groups.map((group) => {
```
Comment (alexkim205) [inline frontend/src/scenes/cohorts/cohortLogic.ts:35]: nit: can we do this in one loop instead of two?

**Review by alexkim205 — APPROVED** [review]
Just a few comments but looks like most of the feedbacks already been given. Looks great :)

File: posthog/models/filters/mixins/property.py (hunk 6 above)
Comment (liyiy) [reply posthog/models/filters/mixins/property.py:107]: I actually missed the suggestion itself because there were so many comments, thanks for the reminder lol

Comment (liyiy) [reply posthog/models/filters/mixins/property.py:107]: hmm just tried it out and it doesn't work (results aren't changing at all), I'll take a closer look at cleaning this up in the back in a follow up PR

File: frontend/src/scenes/cohorts/cohortLogic.ts (hunk 13 above)
Comment (liyiy) [reply frontend/src/scenes/cohorts/cohortLogic.ts:35]: yea I just didn't want to touch existing code too much, will consider for the next PR
